org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/UiPlugin.java

```diff
@@ -127,17 +128,30 @@ private void registerListeners() {
         CorePlugin.listenerRegistry().addEventListener(this.workingSetsAddingProjectCreatedListener);
 
         this.contextActivatingSelectionListener = new ContextActivatingSelectionListener(UiPluginConstants.GRADLE_NATURE_CONTEXT_ID, Predicates.hasGradleNature(), getWorkbench());
-        ((ISelectionService) getWorkbench().getActiveWorkbenchWindow().getService(ISelectionService.class)).addSelectionListener(this.contextActivatingSelectionListener);
+        
+        Display.getDefault().asyncExec(new Runnable() {
+
+                @Override
+                public void run() {
+                    ((ISelectionService) getWorkbench().getActiveWorkbenchWindow().getService(ISelectionService.class)).addSelectionListener(UiPlugin.this.contextActivatingSelectionListener);
+                }
+        });
     }
 
     @SuppressWarnings({"cast", "RedundantCast"})
     private void unregisterListeners() {
-        // if the selection service is disposed, then the listeners are already removed
-        // (see the javadoc on ISelectionService.addSelectionListener(ISelectionListener listener))
-        ISelectionService selectionService = ((ISelectionService) getWorkbench().getService(ISelectionService.class));
-        if (selectionService != null) {
-            selectionService.removeSelectionListener(this.contextActivatingSelectionListener);
-        }
+        Display.getDefault().asyncExec(new Runnable() {
```

Comment:
The Display object should be obtained through the IWorkbench interface and should be tested if it was already disposed. IMHO it should look like this:

Display display = PlatformUI.getWorkbench().getDisplay();
if (display != null && !display.isDisposed()) {
    display.asyncExec(...)
}